### ISSUE TITLE
Help with BIOMOD_4.2-6-1 - [binary outputs ¿how can I change the fitting of the model?]

### ISSUE THREAD

**Issue by zali37** [biomodhub/biomod2#552]
Hello biomod team. There are some issues that I do not understand quite well.

**Context and question**
I made a model with the ANN algorithm. In the argument em.algo of Biomod_EnsembleModeling I put EMmean, EMcv, and EMca and I think EMmean current prediction is ok, but I also want to inform the binary EMca prediction but I think it is overfitting. I wonder if there is a way to change the threshold of that conversion.

**Code used**
darwinii_format <- BIOMOD_FormatingData(
  resp.name = "darwinii",
  resp.var = darwinii$presence,
  resp.xy = darwinii[, c("lon", "lat")],
  expl.var = stk_current,
  eval.resp.var = darwinii_eval$presence,
  eval.expl.var = stk_current,
  eval.resp.xy = darwinii_eval[, c("lon", "lat")],
  filter.raster = T,
  na.rm = F
)
-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.formated.data -=-=-=-=-=-=-=-=-=-=-=-=-=

dir.name =  .

sp.name =  darwinii

	 24 presences,  112 true absences and  0 undefined points in dataset

	 7 explanatory variables

   clip_bio3       clip_bio6        clip_bio8        clip_bio9     
 Min.   :34.92   Min.   :-8.492   Min.   :-2.930   Min.   :-4.233  
 1st Qu.:45.19   1st Qu.:-2.878   1st Qu.: 2.784   1st Qu.: 3.529  
 Median :47.17   Median :-1.502   Median : 6.643   Median : 7.014  
 Mean   :45.95   Mean   :-1.425   Mean   : 6.100   Mean   : 7.902  
 3rd Qu.:48.02   3rd Qu.: 0.151   3rd Qu.: 7.911   3rd Qu.:11.944  
 Max.   :51.54   Max.   : 2.884   Max.   :14.975   Max.   :19.176  
   clip_bio13      clip_bio15     pedregal_total
 Min.   : 16.0   Min.   : 8.682   Min.   :1.0   
 1st Qu.: 33.0   1st Qu.:11.937   1st Qu.:1.0   
 Median :141.5   Median :17.242   Median :1.0   
 Mean   :153.2   Mean   :21.390   Mean   :1.5   
 3rd Qu.:238.0   3rd Qu.:26.370   3rd Qu.:1.0   
 Max.   :579.0   Max.   :60.632   Max.   :5.0   

Evaluation data :

	 8 presences,  35 true absences and  0 undefined points in dataset

   clip_bio3       clip_bio6          clip_bio8         clip_bio9     
 Min.   :34.35   Min.   :-6.62000   Min.   :-0.8413   Min.   :-2.464  
 1st Qu.:44.84   1st Qu.:-3.01400   1st Qu.: 3.0625   1st Qu.: 4.595  
 Median :46.46   Median :-1.49600   Median : 4.5587   Median : 8.385  
 Mean   :45.16   Mean   :-1.63355   Mean   : 5.4681   Mean   : 8.564  
 3rd Qu.:48.10   3rd Qu.: 0.02737   3rd Qu.: 7.5867   3rd Qu.:12.279  
 Max.   :50.10   Max.   : 2.58000   Max.   :13.9833   Max.   :18.825  
   clip_bio13      clip_bio15     pedregal_total 
 Min.   : 18.0   Min.   : 8.293   Min.   :1.000  
 1st Qu.: 27.0   1st Qu.:11.816   1st Qu.:1.000  
 Median :101.0   Median :23.561   Median :1.000  
 Mean   :128.3   Mean   :23.884   Mean   :1.558  
 3rd Qu.:235.0   3rd Qu.:31.186   3rd Qu.:1.000  
 Max.   :382.0   Max.   :52.647   Max.   :5.000  

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
```
darwinii_modelo_ANN <- BIOMOD_Modeling(
  bm.format = darwinii_format,
  modeling.id = "solo_darwinii",
  models = "ANN",
  CV.strategy = "random",
  CV.nb.rep = 5,
  CV.perc = 0.75,
  var.import = 3,
  OPT.strategy = "user.defined",
  OPT.user = darwinii_opt_ANN,
  metric.eval = c("KAPPA","TSS","ROC")
)

darwinii_ensemble_models_ANN <- BIOMOD_EnsembleModeling(
  bm.mod = darwinii_modelo_ANN,
  models.chosen = "all",
  em.by = "all",
  em.algo = c("EMmean", "EMca", "EMcv"),
  metric.select = c("TSS","ROC"),
  metric.select.thresh = c(0.7,0.7),
  metric.eval = c("TSS","ROC"),
  var.import = 3
)

darwinii_proj_ANN <- BIOMOD_Projection(
  bm.mod = darwinii_modelo_ANN,
  proj.name = "darwinii",
  new.env = stk_current,
  models.chosen = "all",
  metric.binary = "TSS",
  build.clamping.mask = FALSE,
  do.stack = FALSE,
  output.format = ".img")

darwinii_ens_proj_ANN <- BIOMOD_EnsembleForecasting(
  bm.em = darwinii_ensemble_models_ANN,
  bm.proj = darwinii_proj_ANN,
  proj.name = "ENMdarwinii_current",
  models.chosen = 'all',
  metric.binary = c('TSS',"ROC"),
  metric.filter = 'all',
  do.stack = FALSE,
  output.format = ".img")

```
**Other outputs** _(optional)_ 
![image](https://github.com/user-attachments/assets/3846da4a-b112-4e49-b956-dd2b0773f0cb)

These are the output maps. On the left is the EMmean, in the middle is the EMcv, and on the right is the binary projection EMca that I think is overestimating the probability of ocurrence (1) .

**Environment Information**
> sessionInfo()
R version 4.3.3 (2024-02-29 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.utf8 
[2] LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

time zone: America/Buenos_Aires
tzcode source: internal

attached base packages:
[1] parallel  splines   stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] caret_6.0-94         ggplot2_3.5.1        snowfall_1.84-6.3   
 [4] snow_0.4-4           biomod2_4.2-6-1      xgboost_1.7.8.1     
 [7] randomForest_4.7-1.2 maxnet_0.1.4         earth_5.3.3         
[10] plotmo_3.6.4         plotrix_3.8-4        Formula_1.2-5       
[13] gbm_2.2.2            lattice_0.22-5       doParallel_1.0.17   
[16] iterators_1.0.14     ggtext_0.1.2         tidyterra_0.6.1     
[19] mgcv_1.9-1           nlme_3.1-166         gam_1.22-5          
[22] foreach_1.5.2        raster_3.6-26        sp_2.1-4            
[25] mda_0.5-4            class_7.3-22         rpart_4.1.23        
[28] nnet_7.3-19          terra_1.7-78        

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1       timeDate_4032.109      farver_2.1.2          
 [4] dplyr_1.1.4            reshape_0.8.9          pROC_1.18.5           
 [7] digest_0.6.35          timechange_0.3.0       lifecycle_1.0.4       
[10] sf_1.0-16              survival_3.6-4         magrittr_2.0.3        
[13] compiler_4.3.3         rlang_1.1.3            tools_4.3.3           
[16] utf8_1.2.4             data.table_1.15.4      labeling_0.4.3        
[19] classInt_0.4-10        PresenceAbsence_1.1.11 plyr_1.8.9            
[22] xml2_1.3.6             abind_1.4-5            KernSmooth_2.23-22    
[25] withr_3.0.1            purrr_1.0.2            stats4_4.3.3          
[28] grid_4.3.3             fansi_1.0.6            future_1.34.0         
[31] e1071_1.7-13           colorspace_2.1-1       globals_0.16.3        
[34] scales_1.3.0           MASS_7.3-60.0.1        cli_3.6.2             
[37] generics_0.1.3         rstudioapi_0.16.0      future.apply_1.11.2   
[40] reshape2_1.4.4         commonmark_1.9.1       DBI_1.2.2             
[43] proxy_0.4-27           stringr_1.5.1          vctrs_0.6.5           
[46] hardhat_1.4.0          Matrix_1.6-1.1         jsonlite_1.8.8        
[49] listenv_0.9.1          gower_1.0.1            tidyr_1.3.1           
[52] units_0.8-5            recipes_1.1.0          parallelly_1.38.0     
[55] glue_1.7.0             codetools_0.2-19       dismo_1.3-14          
[58] lubridate_1.9.3        stringi_1.8.4          gtable_0.3.5          
[61] munsell_0.5.1          tibble_3.2.1           pillar_1.9.0          
[64] ipred_0.9-15           lava_1.8.0             R6_2.5.1              
[67] markdown_1.13          gridtext_0.1.5         Rcpp_1.0.12           
[70] prodlim_2024.06.25     xfun_0.46              ModelMetrics_1.2.2.2  
[73] pkgconfig_2.0.3  
```
Thanks a lot for your help.

**Comment by HeleneBlt** [comment]
Hi Zali, 

Thanks for this well-explained issue ! 

To transform the data into binary data, we use the threshold found by optimizing each metric. You can find it in the column `cutoff` of : 
`get_evaluations(darwinii_ensemble_models_ANN)`. 
You can choose which metric you want with the argument `metric.binary` into `BIOMOD_EnsembleForcasting`. 

If you want to define your own threshold, you can recover the probabilities data of your CA model and transform them yourself with [bm_BinaryTransformation](https://biomodhub.github.io/biomod2/reference/bm_BinaryTransformation.html). 

For example : 
```
prob_EMca_TSS <- get_predictions(darwinii_ens_proj_ANN, full.name =  "darwinii_EMcaByTSS_mergedData_mergedRun_mergedAlgo")
plot(prob_EMca_TSS)

binary_EMca_TSS <- bm_BinaryTransformation(prob_CA_TSS, threshold = 600)
plot(binary_EMca_TSS)
```

Hope it helps ! 
Hélène

**Comment by zali37** [comment]
Hello Helene thank you for your help.
Now that I can see the evaluation table, I do not have validation values. I thought it was calculated when I did cross-validation so I do not know why it has only NAs. Can you please clarify this for me? I must be doing something wrong. I will share with you the modeling setting for your knowledge.

cv.k <- bm_CrossValidation(bm.format = darwinii_format,
                           strategy = "kfold",
                           k = 5,
                           nb.rep = 10)

opt.d <- bm_ModelingOptions(data.type = 'binary',
                            models = c("SRE","RF","GAM.gam.gam","ANN"),
                            strategy = "default",
                            bm.format = darwinii_format,
                            calib.lines = cv.k)  

tuned.ann <- bm_Tuning (model = "ANN",
                       tuning.fun = "avNNet",
                       do.formula = T,
                       bm.options = opt.d@options$ANN.binary.nnet.nnet,
                       bm.format = darwinii_format) 

darwinii_opt_ANN <- bm_ModelingOptions(
  data.type = "binary",
  models="ANN",
  bm.format = darwinii_format,
  strategy = "user.defined",
  user.base = "bigboss",
  user.val = list(ANN.binary.nnet.nnet = tuned.ann))

darwinii_modelo_ANN <- BIOMOD_Modeling(
  bm.format = darwinii_format,
  modeling.id = "solo_darwinii",
  models = "ANN",
  CV.strategy = "random",
  CV.nb.rep = 5,
  CV.perc = 0.75,
  var.import = 3,
  OPT.strategy = "user.defined",
  OPT.user = darwinii_opt_ANN,
  metric.eval = c("KAPPA","TSS","ROC"))

darwinii_ensemble_models_ANN <- BIOMOD_EnsembleModeling(
  bm.mod = darwinii_modelo_ANN,
  models.chosen = "all",
  em.by = "all",
  em.algo = c("EMmean", "EMca", "EMcv"),
  metric.select = "TSS",
  metric.select.thresh = 0.7,
  metric.eval = "TSS",
  var.import = 3)

Then I got this table 
> cutoff <- get_evaluations(darwinii_ensemble_models_ANN)
> cutoff
                                             full.name merged.by.PA merged.by.run merged.by.algo filtered.by   algo metric.eval
1 darwinii_EMmeanByTSS_mergedData_mergedRun_mergedAlgo   mergedData     mergedRun     mergedAlgo         TSS EMmean         TSS
2   darwinii_EMcaByTSS_mergedData_mergedRun_mergedAlgo   mergedData     mergedRun     mergedAlgo         TSS   EMca         TSS
  cutoff sensitivity specificity calibration validation evaluation
1  227.0         100      91.071       0.911         NA      0.618
2  247.5         100      91.964       0.920         NA      0.550

Thanks a lot
Natalia

**Comment by HeleneBlt** [comment]
Hi Natalia ! 

There is actually no error here 🙂

- At the single models step, you get one value for calibration and one for validation. And if you look at your single models evaluation table, you will have a value for calibration and for validation : 
`get_evaluations(darwinii_modelo_ANN)`

- But when going to the ensemble model step, you ask to merge all models together, meaning that all repetitions of all PA datasets are combined together. Therefore at this step, calibration and validation lines do not match anymore, and only one evaluation value is calculated over the whole dataset.

However, you can change this with the parameter `em.by` in `BIOMOD_EnsembleModeling`. For example, you can choose `"PA+run"` and you will have one model for each PA+run combination. So you can have a validation dataset. But you will not have one unique ensemble model. 

I hope it's clearer that way ! 
Hélène

**Comment by zali37** [comment]
Hi Helene
Thank you so much for your explanations, now I understand better.
Take care!
